### PR TITLE
Functions

### PR DESCRIPTION
**Pull request by jmdobry**

### CODE REVIEW

**Comment by jmdobry** [comment]
@ace-n @jasonpolites PTAL

**Comment by ace-n** [comment]
@jmdobry Missing module - mind fixing that?

**Comment by jmdobry** [comment]
Will do

File: functions/pubsub/index.js [hunk 1]
```diff
@@ -1,83 +1,88 @@
...
+ * Triggered from a message on a Cloud Pub/Sub topic.
  *
- * @param {Object} context Cloud Function context.
- * @param {Function} context.success Success callback.
- * @param {Function} context.failure Failure callback.
- * @param {Object} data Request data, in this case an object provided by the Pub/Sub trigger.
- * @param {Object} data.message Message that was published via Pub/Sub.
+ * @param {object} event The Cloud Functions event.
+ * @param {object} event.data The Cloud Pub/Sub Message object.
+ * @param {string} event.data.data The "data" property of the Cloud Pub/Sub Message.
```
Comment (jmdobry) [inline functions/pubsub/index.js]: That's what Cloud Functions implemented. There's literally no way around that.

File: functions/background/index.js [hunk 2]
```diff
@@ -1,68 +1,72 @@
...
- * a "context" argument to the function.
+ * a "callback" argument to the function.
  *
- * @param {Object} data Request data, provided by a trigger.
+ * @param {object} event The Cloud Functions event.
+ * @param {object} event.data The event data.
  * @returns {Promise}
  */
-exports.helloPromise = function helloPromise (data) {
+exports.helloPromise = function helloPromise (event) {
```
Comment (ace-n) [inline functions/background/index.js]: Do we need the function names here?

File: functions/datastore/README.md [hunk 3]
```diff
@@ -20,46 +21,75 @@ Functions for your project.
...
+    * Replace `YOUR_BUCKET_NAME` with the name of your Cloud Storage Bucket.
 
-1. Deploy the "ds-del" function with an HTTP trigger:
+1. Deploy the "del" function with an HTTP trigger:
 
-        gcloud alpha functions deploy ds-del --bucket [YOUR_BUCKET_NAME] --trigger-http --entry-point del
+        gcloud alpha functions deploy del --stage-bucket YOUR_BUCKET_NAME --trigger-http
 
-    * Replace `[YOUR_BUCKET_NAME]` with the name of your Cloud Storage Bucket.
+    * Replace `YOUR_BUCKET_NAME` with the name of your Cloud Storage Bucket.
```
Comment (ace-n) [inline functions/datastore/README.md]: Does this need to be repeated every time?

File: functions/errorreporting/index.js [hunk 4]
```diff
@@ -36,106 +36,114 @@ function reportError (err, callback) {
...
- * @param {Object} context Cloud Function context object.
- * @param {Object} data Request data, provided by a trigger.
- * @param {string} data.message Message, provided by the trigger.
+ * @param {object} event The Cloud Functions event.
+ * @param {object} event.data The event data.
+ * @param {string} event.data.message Message, provided by the trigger.
+ * @param {function} The callback function.
  */
-exports.helloBackgroundError = function helloBackgroundError (context, data) {
+exports.helloBackgroundError = function helloBackgroundError (event, callback) {
```
Comment (jmdobry) [inline functions/errorreporting/index.js]: The comment is supposed to be there, indicating that the user would put their own code there.

File: functions/errorreporting/index.js [hunk 5]
```diff
@@ -36,106 +36,114 @@ function reportError (err, callback) {
...
   try {
-    if (!data.message) {
+    if (!event.data.message) {
       throw new Error('"message" is required!');
     }
-    // All is good, respond with a message
-    return context.success('Hello World!');
+
+    // Do something
```
Comment (jmdobry) [inline functions/errorreporting/index.js]: The comment is supposed to be there, indicating that the user would put their own code there.

File: functions/datastore/test/index.test.js [hunk 6]
```diff
@@ -1,310 +1,266 @@
...
 
-function getMockContext () {
-  return {
-    success: sinon.stub(),
-    failure: sinon.stub()
-  };
-}
+describe(`functions:datastore`, () => {
+  it(`set: Set fails without a value`, () => {
+    const expectedMsg = `Value not provided. Make sure you have a "value" property in your request`;
```
Comment (jmdobry) [inline functions/datastore/test/index.test.js]: I feel like the general pattern I follow is to pull the expected value into a variable when the line is getting long.

File: functions/gcs/index.js [hunk 7]
```diff
@@ -1,70 +1,75 @@
...
+// Instantiates a client
+const storage = Storage();
+// [END functions_word_count_setup]
 
-function getFileStream (bucketName, fileName) {
-  if (!bucketName) {
-    throw new Error('Bucket not provided. Make sure you have a ' +
-      '"bucket" property in your request');
+// [START functions_word_count_stream]
+function getFileStream (file) {
```
Comment (jmdobry) [inline functions/gcs/index.js]: I'll delete the region tag, as it's not shown in the docs.

File: functions/gcs/test/index.test.js [hunk 8]
```diff
@@ -1,38 +1,39 @@
...
+const filename = `sample.txt`;
 
 function getSample () {
-  var file = {
-    createReadStream: function () {
-      var filepath = path.join(__dirname, '../sample.txt');
-      return fs.createReadStream(filepath, { encoding: 'utf8' });
-    }
+  const file = {
+    createReadStream: () => fs.createReadStream(path.join(__dirname, `../${filename}`), { encoding: `utf8` })
```
Comment (jmdobry) [inline functions/gcs/test/index.test.js]: Sure.

File: functions/helloworld/index.js [hunk 9]
```diff
@@ -36,54 +38,105 @@ exports.helloWorld = function helloWorld (context, data) {
...
- * @param {Object} data Request data, provided by a Cloud Storage trigger.
+ * @param {object} event The Cloud Functions event.
+ * @param {function} The callback function.
+ */
+exports.helloGCS = function helloGCS (event, callback) {
+  const file = event.data;
+  const isDelete = file.resourceState === 'not_exists';
+
+  if (isDelete) {
+    console.log(`File ${file.name} deleted.`);
```
Comment (jmdobry) [inline functions/helloworld/index.js]: No, as in this sample the idea is that the user themselves would perform different actions based on the value of `isDelete`. Smashing the conditional down into a string eliminates that.

File: functions/helloworld/test/index.test.js [hunk 10]
```diff
@@ -1,141 +1,187 @@
...
+    }, Error, expectedMsg);
+  });
+
+  it(`helloError2: should throw an error`, () => {
+    assert.throws(() => {
+      program.helloError2();
+    });
+  });
+
+  it(`helloError3: should throw an error`, () => {
```
Comment (jmdobry) [inline functions/helloworld/test/index.test.js]: Fixed

File: functions/ocr/app/index.js [hunk 11]
```diff
@@ -108,168 +93,140 @@ function detectText (filename, image, callback) {
...
+      }
 
-    var bucket = storage.bucket(data.bucket);
-    var file = bucket.file(data.name);
-    detectText(data.name, file, function (err) {
-      if (err) {
-        console.error(err);
-        return context.failure(err);
+      if (!file.bucket) {
+        throw new Error('Bucket not provided. Make sure you have a "bucket" property in your request');
```
Comment (jmdobry) [inline functions/ocr/app/index.js]: Too much abstraction.

File: functions/ocr/app/test/index.test.js [hunk 12]
```diff
@@ -1,438 +1,271 @@
...
-      failure: function () {
-        assert.equal(context.failure.calledOnce, true);
-        assert.equal(context.failure.firstCall.args[0], expectedMsg);
-        assert.equal(context.success.called, false);
-        assert.equal(console.error.calledWith(expectedMsg), true);
-        done();
+  it(`processImage processes an image`, () => {
+    const event = {
+      data: {
+        bucket: bucketName,
```
Comment (jmdobry) [inline functions/ocr/app/test/index.test.js]: No, as I want it to be clear that the variable represents a string. `bucket` is the variable name I would assign for an actual bucket object.

File: functions/ocr/app/test/index.test.js [hunk 13]
```diff
@@ -1,438 +1,271 @@
...
-        assert.equal(context.failure.calledOnce, true);
-        assert.equal(context.failure.firstCall.args[0], expectedMsg);
-        assert.equal(context.success.called, false);
-        assert.equal(console.error.calledWith(expectedMsg), true);
-        done();
+  it(`translateText fails without text`, () => {
+    const error = new Error(`Text not provided. Make sure you have a "text" property in your request`);
+    const event = {
+      data: {
+        data: Buffer.from(JSON.stringify({})).toString(`base64`)
```
Comment (jmdobry) [inline functions/ocr/app/test/index.test.js]: I want it to be clear that you can use an actual object when constructing the message payload, and that the object must be stringifed.

File: functions/pubsub/test/index.test.js [hunk 14]
```diff
@@ -1,141 +1,137 @@
...
+    delete sample.mocks.req.body.topic;
+    sample.program.publish(sample.mocks.req, sample.mocks.res);
 
-    assert.equal(context.failure.calledOnce, true);
-    assert.equal(context.failure.firstCall.args[0], expectedMsg);
-    assert.equal(context.success.called, false);
+    assert.deepEqual(sample.mocks.res.status.callCount, 1);
+    assert.deepEqual(sample.mocks.res.status.firstCall.args, [500]);
+    assert.deepEqual(sample.mocks.res.send.callCount, 1);
+    assert.deepEqual(sample.mocks.res.send.firstCall.args[0].message, expectedMsg);
```
Comment (jmdobry) [inline functions/pubsub/test/index.test.js]: Fixed

File: functions/pubsub/test/index.test.js [hunk 15]
```diff
@@ -1,141 +1,137 @@
...
+    delete sample.mocks.req.body.message;
+    sample.program.publish(sample.mocks.req, sample.mocks.res);
 
-    assert.equal(context.failure.calledOnce, true);
-    assert.equal(context.failure.firstCall.args[0], expectedMsg);
-    assert.equal(context.success.called, false);
+    assert.deepEqual(sample.mocks.res.status.callCount, 1);
+    assert.deepEqual(sample.mocks.res.status.firstCall.args, [500]);
+    assert.deepEqual(sample.mocks.res.send.callCount, 1);
+    assert.deepEqual(sample.mocks.res.send.firstCall.args[0].message, expectedMsg);
```
Comment (jmdobry) [inline functions/pubsub/test/index.test.js]: Fixed

File: functions/sendgrid/index.js [hunk 16]
```diff
@@ -1,101 +1,105 @@
...
  * @param {string} data.subject Email subject line.
  * @param {string} data.body Body of the email subject line.
- * @returns {Object} Payload object.
+ * @returns {object} Payload object.
  */
 function getPayload (requestBody) {
   if (!requestBody.to) {
-    var error = new Error('To email address not provided. Make sure you have a ' +
-      '"to" property in your request');
+    const error = new Error('To email address not provided. Make sure you have a "to" property in your request');
```
Comment (jmdobry) [inline functions/sendgrid/index.js]: 1) too much abstraction
2) I have to first save the error to a variable so I can add the `code` property to it.

File: functions/sendgrid/index.js [hunk 17]
```diff
@@ -189,173 +200,129 @@ function fixNames (obj) {
...
-
-    if (!data.bucket) {
-      throw new Error('Bucket not provided. Make sure you have a ' +
-        '"bucket" property in your request');
-    }
-    if (!data.name) {
-      throw new Error('Filename not provided. Make sure you have a ' +
-        '"name" property in your request');
-    }
+exports.sendgridLoad = function sendgridLoad (event) {
```
Comment (jmdobry) [inline functions/sendgrid/index.js]: Fixed

File: functions/background/index.js (hunk 2 above)
Comment (jmdobry) [inline functions/background/index.js]: Makes for better stack traces.

File: functions/datastore/README.md (hunk 3 above)
Comment (jmdobry) [inline functions/datastore/README.md]: Done.

**Comment by jmdobry** [comment]
Fixed some stuff related to the OCR sample.

**Pull request merged**